From b4253e66ecc84249a8127724decaaaed52c747a7 Mon Sep 17 00:00:00 2001
From: Fatih Uzunoglu <fuzun54@outlook.com>
Date: Thu, 15 Aug 2024 16:14:12 +0300
Subject: [PATCH] Do not link D3D9

On some configurations `d3d9.dll` might not be present,
such as Windows 11 PE, reportedly.
---
 src/plugins/platforms/direct2d/CMakeLists.txt          | 1 -
 src/plugins/platforms/windows/CMakeLists.txt           | 1 -
 src/plugins/platforms/windows/qwindowsopengltester.cpp | 5 ++++-
 3 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/plugins/platforms/direct2d/CMakeLists.txt b/src/plugins/platforms/direct2d/CMakeLists.txt
index e3b8204ae45..1550b0d136f 100644
--- a/src/plugins/platforms/direct2d/CMakeLists.txt
+++ b/src/plugins/platforms/direct2d/CMakeLists.txt
@@ -79,7 +79,6 @@ qt_internal_add_plugin(QWindowsDirect2DIntegrationPlugin
         winspool
         wtsapi32
         comdlg32
-        d3d9
         mscms
 )
 
diff --git a/src/plugins/platforms/windows/CMakeLists.txt b/src/plugins/platforms/windows/CMakeLists.txt
index 376915d4b51..b3340f4f035 100644
--- a/src/plugins/platforms/windows/CMakeLists.txt
+++ b/src/plugins/platforms/windows/CMakeLists.txt
@@ -66,7 +66,6 @@ qt_internal_add_plugin(QWindowsIntegrationPlugin
         winspool
         wtsapi32
         comdlg32
-        d3d9
         mscms
 )
 
diff --git a/src/plugins/platforms/windows/qwindowsopengltester.cpp b/src/plugins/platforms/windows/qwindowsopengltester.cpp
index aed79e11978..cfc7b53696c 100644
--- a/src/plugins/platforms/windows/qwindowsopengltester.cpp
+++ b/src/plugins/platforms/windows/qwindowsopengltester.cpp
@@ -67,7 +67,10 @@ private:
 QDirect3D9Handle::QDirect3D9Handle()
 {
 #ifndef QT_NO_OPENGL
-    m_direct3D9 = Direct3DCreate9(D3D_SDK_VERSION);
+    QSystemLibrary d3d9dll(QStringLiteral("d3d9"));
+    typedef IDirect3D9* (WINAPI *Direct3DCreate9Func)(UINT);
+    if (const auto direct3DCreate9 = (Direct3DCreate9Func)(d3d9dll.resolve("Direct3DCreate9")))
+        m_direct3D9 = direct3DCreate9(D3D_SDK_VERSION);
 #endif
 }
 
-- 
2.55.0

